### PR TITLE
[Feature] add sparse embedding.

### PR DESCRIPTION
**Pull request by zheng-da**
## Description
This is to add distributed sparse embedding support for distributed GNN training.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

File: python/dgl/distributed/sparse_emb.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    def __call__(self, idx):
+        emb = F.attach_grad(self._tensor[idx])
+        self._trace.append((idx, emb))
+        return emb
+
+def sparse_adagrad_optimize(data_store, name, indices, data):
+    ''' Update the embeddings with sparse Adagrad.
+
+    This function runs on the KVStore server. It updates the gradients by scaling them
+    according to the state sum. The gradients have been adjusted by the learning rate
```
Comment (classicsong) [inline python/dgl/distributed/sparse_emb.py:47]: Will this cause any problem? As the state_sum is calculated by the values adjusted by the learning rate. This is different from the SOTA impl.

Comment (zheng-da) [reply python/dgl/distributed/sparse_emb.py:47]: you are right. it should be adjusted by the learning rate beforehand.

File: python/dgl/distributed/sparse_emb.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        The name of the embeddings
+    shape : tuple of int
+        The shape of the embedding. The first dimension should be the number of nodes.
+    initializer : callable
+        The function to create the initial data.
+    '''
+    def __init__(self, g, name, shape, initializer):
+        assert shape[0] == g.number_of_nodes()
+        g._client.init_data(_get_ndata_name(name), shape, F.float32, 'node',
+                            g.get_partition_book(), initializer)
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply python/dgl/distributed/sparse_emb.py:31]: i think we need a separate implementation for non-distributed version.

File: python/dgl/backend/backend.py [hunk 3]
```diff
@@ -718,6 +733,23 @@ def scatter_row(data, row_index, value):
     """
     pass
 
+def index_add_inplace(data, row_idx, value):
+    """Add the values into the data tensor using the row index inplace.
+
+    If two row indices are the same, the corresponding values are sum up before
+    adding to the data tensor.
```
Comment (classicsong) [inline python/dgl/backend/backend.py:740]: We can add an example here as:
```
arr = th.zeros((10))
F. index_add_inplace(arr, th.tensor([0, 1, 1]), th.tensor([1.0, 1.0, 1.0]))                                                                         
tensor([1., 2., 0., 0., 0., 0., 0., 0., 0., 0.])
```

File: python/dgl/distributed/__init__.py [hunk 4]
```diff
@@ -1,8 +1,9 @@
 """DGL distributed."""
 
-from .dist_graph import DistGraphServer, DistGraph, node_split, edge_split
+from .dist_graph import DistGraphServer, DistGraph, DistTensor, node_split, edge_split
 from .partition import partition_graph, load_partition, load_partition_book
 from .graph_partition_book import GraphPartitionBook, RangePartitionBook, PartitionPolicy
+from .sparse_emb import SparseAdagrad, SparseNodeEmbedding
```
Comment (classicsong) [inline python/dgl/distributed/__init__.py:6]: SparseNodeEmbedding can be SparseEmbedding. There is no difference between SparseNodeEmbedding and SparseEdgeEmbedding.

File: python/dgl/distributed/sparse_emb.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        embs = data_store[name]
+        state_sum = data_store[name + "_sum"]
+        with F.no_grad():
+            grad_sum = F.mean(grad_values * grad_values, 1)
+            F.index_add_inplace(state_sum, grad_indices, grad_sum)
+            std = state_sum[grad_indices]  # _sparse_mask
+            std_values = F.unsqueeze((F.sqrt(std) + 1e-10), 1)
+            F.index_add_inplace(embs, grad_indices, grad_values / std_values * (-self._lr))
+
+def _init_state(shape, dtype):
```
Comment (classicsong) [inline python/dgl/distributed/sparse_emb.py:72]: inside SparseAdagradUDF class？

File: python/dgl/distributed/__init__.py (hunk 4 above)
Comment (zheng-da) [reply python/dgl/distributed/__init__.py:6]: there is. It gets a graph as input and creates embeddings for nodes.

File: python/dgl/distributed/sparse_emb.py (hunk 5 above)
Comment (zheng-da) [reply python/dgl/distributed/sparse_emb.py:72]: why inside this class?

**Pull request merged**